testplan/runnable/interactive/base.py

```diff
@@ -856,32 +740,18 @@ def _log_async_exceptions(self, future):
         except Exception:
             self.logger.exception("Exception caught in async function")
 
-    def _create_testplan_report(self, test_reports):
-        """Create a single testplan report from many test reports."""
-        report = testplan.report.TestReport(
-            name=self.cfg.name, uid=self.cfg.name
-        )
-
-        for test_report in test_reports:
-            report.append(test_report)
-
-        return report
-
-    def _merge_test_reports(self, results, test_uid, runner_uid):
-        """Merge all test reports from a test run into one."""
-        test_reports = [
-            res
-            for res in results
-            if isinstance(res, testplan.report.TestGroupReport)
-        ]
-
-        if not test_reports:
-            return None
-
-        test = self.test(test_uid, runner_uid=runner_uid)
-        initial_report = test.dry_run().report
+    def _merge_testcase_reports(self, testcase_reports):
+        """Merge all test reports from a test run into our report."""
+        with self.report_mutex:
+            for report, parent_uids in testcase_reports:
+                self.logger.debug(
+                    "Merging testcase report %s with parent UIDs %s",
+                    report,
+                    testcase_reports,
+                )
 
-        for report in test_reports:
-            initial_report.merge(report)
+                parent_entry = self.report
+                for uid in parent_uids:
+                    parent_entry = parent_entry[uid]
 
-        return initial_report
+                parent_entry[report.uid] = report
```

Comment:
Why not `return` or `yield` the report after this line?

Reply:
what benefit would that give?

Yielding a list of testcase reports probably isn't very useful, if they all belong to different MultiTests / testsuites

We _could_ return the root self.report object here. However that would be misleading IMO, as it will include the results of all testcases that have been run, not just those run in the current run.

Overall, I think that returning a report _and also_ mutating the self.report as a side-effect is a potentially dangerous pattern. By returning a report, it sorta gives the impression of being a pure function that runs tests and returns the results of only those tests as a single report tree - but that's not what is actually happening. If there is no return type, it is obvious that there must be some side-effect, which in this case is mutating the report tree at self.report.

Reply:
It is a cost-free way to minimize API breakage.

Reply:
There is no API breakage - this is brand new code I added just a week or so ago, nothing apart from the unit test I modified is using it.

Reply:
^ this is new information to me and nullifies my concerns about API breakage.